src/fsharp/ConstraintSolver.fs

```diff
@@ -1246,8 +1246,12 @@ and SolveMemberConstraint (csenv:ConstraintSolverEnv) permitWeakResolution ndeep
                           let objtys = minfo.GetObjArgTypes(amap, m, minst)
                           Some(CalledMeth<Expr>(csenv.InfoReader,None,false,FreshenMethInfo,m,AccessibleFromEverywhere,minfo,minst,minst,None,objtys,[(callerArgs,[])],false,false,None)))
 
-              let methOverloadResult,errors = 
-                  CollectThenUndo (fun trace -> ResolveOverloading csenv (WithTrace(trace)) nm ndeep (Some traitInfo) (0,0) AccessibleFromEverywhere calledMethGroup false (Some rty))  
+              let methOverloadTrace = Trace.New()
+              let methOverloadResult,errors = ResolveOverloading csenv (WithTrace methOverloadTrace) nm ndeep (Some traitInfo) (0,0) AccessibleFromEverywhere calledMethGroup false (Some rty)  
+              match methOverloadResult, trace with
```

Comment:
Could you make this a method on the ``OptionalTrace`` type please (so ``actions`` is effectively private and encapsulated)?